File: backend/app/routers/alerts.py
from bson import ObjectId
from fastapi import APIRouter, Depends, HTTPException

from ..core.database import get_db
from ..core.patient_resolver import resolve_patient_id
from ..models.alerts import AlertOut

router = APIRouter(prefix="/api/alerts", tags=["alerts"])


@router.get("", response_model=list[AlertOut])
async def list_alerts(patient_id: str = Depends(resolve_patient_id)):
    """Get the 20 most recent unrecognized face alerts for this patient."""
    db = get_db()
    query = {"patient_id": patient_id}
    docs = (
        await db["alerts"]
        .find(query)
        .sort("timestamp", -1)
        .limit(20)
        .to_list(length=20)
    )
    return [
        AlertOut(
            id=str(d["_id"]),
            type=d.get("type", "unknown_face"),
            timestamp=d.get("timestamp", ""),
            patient_id=str(d.get("patient_id", patient_id)),
        )
        for d in docs
    ]


@router.delete("/{alert_id}")
async def dismiss_alert(
    alert_id: str, patient_id: str = Depends(resolve_patient_id)
):
    db = get_db()
    result = await db["alerts"].delete_one({"_id": ObjectId(alert_id)})
    if result.deleted_count == 0:
        raise HTTPException(status_code=404, detail="Alert not found")
    return {"status": "ok"}

File: src/server-routes/alerts.ts
import { Router } from "express";
import { ObjectId } from "mongodb";
import { getDb } from "../server-core/database";
import { authMiddleware } from "../server-core/security";
import { resolvePatientId } from "../server-core/patientResolver";

const router = Router();

// GET /api/alerts
router.get("/", authMiddleware, resolvePatientId, async (req, res) => {
  try {
    const db = getDb();
    const docs = await db
      .collection("alerts")
      .find({ patient_id: req.patientId! })
      .sort({ timestamp: -1 })
      .limit(20)
      .toArray();

    res.json(
      docs.map((d) => ({
        id: String(d._id),
        type: d.type || "unknown_face",
        timestamp: d.timestamp || "",
        patient_id: String(d.patient_id || req.patientId),
      }))
    );
  } catch (err) {
    console.error("list alerts error:", err);
    res.status(500).json({ detail: "Internal server error" });
  }
});

// DELETE /api/alerts/:alertId
router.delete("/:alertId", authMiddleware, resolvePatientId, async (req, res) => {
  try {
    const db = getDb();
    const result = await db.collection("alerts").deleteOne({ _id: new ObjectId(req.params.alertId as string), patient_id: req.patientId! });
    if (result.deletedCount === 0) {
      res.status(404).json({ detail: "Alert not found" });
      return;
    }
    res.json({ status: "ok" });
  } catch (err) {
    console.error("dismiss alert error:", err);
    res.status(500).json({ detail: "Internal server error" });
  }
});

export default router;
